designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+exports.config = (context) => {
+
+    // do something
+
+    return {
+        rules: {
+            "semi": ["error", "always"]
+        }
+    };
+
+};
+```
+
+The `context` object has the following members:
+
+* `core` - information about the ESLint core that is using the config
+    * `version` - the version of ESLint being used
+    * `hasRule(ruleId)` - determine if the given rule is in the core
+    * `getConfig(configId)` - returns one of the predefined ESLint configs 
+* `cwd` - the current working directory for ESLint (might be different than `process.cwd()`, see https://github.com/eslint/eslint/issues/11218)
+
+This information allows users to make logical decisions about how the config should be constructed.
+
+#### Checking for Rule Existence
+
+One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:
+
+```js
+const { Config } = require("@eslint/config");
+
+exports.config = (context) => {
+    const myConfig = new Config({
+        ruledefs: {
+            custom: ruleLoader.loadFromDirectory("./custom-rules")
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+
+};
+```
+
+#### Extending `eslint:recommended`
+
+Advanced configs must be used to extend builtin configs such as `eslint:recommended` and `eslint:all`:
+
+```js
+const { Config } = require("@eslint/config");
+
+exports.config = (context) => {
+
+    const recommendedConfig = context.core.getConfig("eslint:recommended");
+
+    return Config.create(recommendedConfig, {
+        ruledefs: {
+            custom: ruleLoader.loadFromDirectory("./custom-rules")
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    });
+};
+```
+
+### Configuration Location Resolution
+
+When ESLint is executed, the following steps are taken to find the `.eslint.js` file to use:
+
+1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
+1. Otherwise:
+    a. Look for `.eslint.js` in the current working directory. If founnd, stop searching and use that file.
+    b. If not found, search up the directory hierarchy looking for `.eslint.js`.
+    c. If a `.eslint.js` file is found at any point, stop searching and use that file.
+    d. If `/` is reached without finding `.eslint.js`, then stop searching and output a "no configuration found" error.
+
+This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `.eslint.js` is found).
```

Comment:
To clarify, this is assuming the `.eslint.js` file doesn't use the provided `cwd` value, right? Or would the CWD be set to the directory of the `.eslint.js` file by default?

Reply:
I meant that the statement "This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory" is not necessarily true in the unusual case where a config file is explicitly changing what config it returns depending on the `cwd` argument, because the `cwd` would be different in the two invocations.

Reply:
Oh yes, good point. Yes, there is a theoretical way that could end up different but I'd expect that to be an edge case. So I think the word "allow" is correct in my statement rather than "enforce."